### PR TITLE
Add the schema to Kinto server while creating the collection.

### PR DESCRIPTION
**Pull request by Natim**
- [x] Grab the last version of the schema from https://github.com/mozilla-services/amo-blocklist-ui
- [x] Create or Update the collection with the appropriate schema if any
- [x] Add a functional test to make sure that current schema can handle xml2kinto importation.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -13,6 +13,10 @@ matrix:
     - python: 3.5
       env:
         - TOX_ENV=py35
+    - env: TOX_ENV=functional
```
Comment (magopian) [inline .travis.yml]: This is a bit strange, shouldn't this `env` be listed with the others above, below line 6? It doesn't need this "include" special matrix hack I believe, as it'll be run with py27 (at least that's what it's doing [at the moment](https://travis-ci.org/mozilla-services/xml2kinto/jobs/116093357))

Comment (magopian) [inline .travis.yml]: Or is it needed to have the `before_install`?

File: Makefile [hunk 2]
```diff
@@ -41,6 +42,9 @@ tests:
 	@rm -fr .coverage
 	$(VENV)/bin/tox
 
+functional: need-kinto-running
```
Comment (magopian) [inline Makefile]: rename to `functional_tests`?

File: Makefile [hunk 3]
```diff
@@ -56,3 +60,16 @@ sync: install
...
+
+install-kinto: $(VENV)/bin/kinto
+$(VENV)/bin/kinto: install
+	$(VENV)/bin/pip install kinto
+
+run-kinto: $(VENV)/bin/kinto
+	$(VENV)/bin/kinto --ini config/kinto.ini start
+
+need-kinto-running:
+	@curl http://localhost:8888/v1 2>/dev/null 1>&2 || (echo "Run 'make run-signer' before starting tests." && exit 1)
```
Comment (magopian) [inline Makefile]: Can you explain what this line is doing? It's confusing, is the "echo" relevant?

File: tox.ini [hunk 4]
```diff
@@ -9,6 +9,9 @@ commands =
 deps = -rdev-requirements.txt
 install_command = pip install --pre {opts} {packages}
 
+[testenv:functional]
+commands = xml2kinto -s http://localhost:8888/v1 -x blocklist.xml -S schemas.json
```
Comment (magopian) [inline tox.ini]: replace by `make functional_tests`?

File: config/kinto.ini [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+kinto.cache_url =
+kinto.permission_backend = cliquet.permission.memory
+kinto.permission_url =
+
+#
+# Auth configuration.
+#
+# http://kinto.readthedocs.org/en/latest/configuration/settings.html#authentication
+#
+kinto.userid_hmac_secret = d32712c82034ec704828ff9038c5926fca11463fb92130dd70d8ae7d593294c5
```
Comment (magopian) [inline config/kinto.ini]: Maybe add a comment explaining what this secret corresponds to?

File: config/kinto.ini [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+# buffer-size = 65535
+# post-buffering = 65535
+# plugin = python
+
+
+#
+# Logging configuration
+#
+
+# kinto.logging_renderer = cliquet.logs.ClassicLogRenderer
```
Comment (magopian) [inline config/kinto.ini]: I'd like all those comments to be removed, they feel useless. Also, maybe this file can be stripped down to only keep the values that are different than the defaults?

File: schemas.json [hunk 7]
```diff
@@ -0,0 +1,435 @@
+{
```
Comment (magopian) [inline schemas.json]: I don't think we want the schemas.json to be tracked in this repository, as this PR is pulling it from the "single source of truth" wich is the amo-blocklist-ui repository.

File: xml2kinto/__main__.py [hunk 8]
```diff
@@ -10,6 +12,10 @@
 # options to move to a config file
 XML_FILE = os.path.abspath(os.path.join(os.path.dirname(__file__),
                                         '..', 'blocklist.xml'))
+
+SCHEMA_FILE = os.path.abspath(os.path.join(os.path.dirname(__file__),
+                                           '..', 'schemas.json'))
```
Comment (magopian) [inline xml2kinto/__main__.py]: We usually don't use `..`, but another call to `os.path.dirname`

File: xml2kinto/__main__.py [hunk 9]
```diff
@@ -137,6 +144,15 @@ def main(args=None):
     parser.add_argument('-x', '--xml-file', help='XML Source file',
                         type=str, default=XML_FILE)
 
+    parser.add_argument('-S', '--schema-file', help='JSON Schemas file',
+                        type=str, default=SCHEMA_FILE)
+
+    parser.add_argument('--no-schema', help='Should we handle schemas',
+                        action="store_true")
```
Comment (magopian) [inline xml2kinto/__main__.py]: Do we want/need this option?

File: xml2kinto/__main__.py [hunk 10]
```diff
@@ -137,6 +144,15 @@ def main(args=None):
...
+
+    parser.add_argument('--with-scrapping', action="store_true",
+                        help='Activate blocklist scrapping on AMO')
```
Comment (magopian) [inline xml2kinto/__main__.py]: Not relevant to this PR, it would be better to have it in another PR

File: xml2kinto/__main__.py [hunk 11]
```diff
@@ -172,7 +194,9 @@ def main(args=None):
             xpath='certItems/*',
             kinto_client=kinto_client,
             bucket=args.cert_bucket,
-            collection=args.cert_collection),
+            collection=args.cert_collection,
+            schema=schemas.get(args.cert_bucket,
+                               schemas.get(CERT_COLLECTION))),
```
Comment (magopian) [inline xml2kinto/__main__.py]: What's the idea behind looking for the bucket name first, and then the collection name?

File: xml2kinto/kinto.py [hunk 12]
```diff
@@ -1,7 +1,18 @@
...
+        collection_data['schema'] = schema
+
+    previous_collection = kinto_client.create_collection(
         collection, bucket, permissions=permissions, if_not_exists=True)
+
+    if schema and ('schema' not in previous_collection or
+                   previous_collection['schema'] != schema):
+        kinto_client.patch_collection(collection=collection, bucket=bucket,
+                                      data=collection_data)
+
```
Comment (magopian) [inline xml2kinto/kinto.py]: So all this is for the use case where we update the schema on an existing collection? If so, please 1/ extract that into a helper function 2/ document it (comments and docstring)

File: xml2kinto/tests/test_kinto.py [hunk 13]
```diff
@@ -23,8 +25,51 @@ def test_get_kinto_records_try_to_create_the_collection_with_permissions():
...
+    get_kinto_records(kinto_client,
+                      mock.sentinel.bucket,
+                      mock.sentinel.collection,
+                      mock.sentinel.permissions,
+                      schema={'foo': 'bar'})
+
+    kinto_client.patch_collection.assert_called_with(
+        bucket=mock.sentinel.bucket,
+        collection=mock.sentinel.collection,
+        data={"schema": {"foo": "bar"}})
```
Comment (magopian) [inline xml2kinto/tests/test_kinto.py]: Do you want to add a test for the case where the collection exists and the schema is the same?

File: xml2kinto/tests/test_main.py [hunk 14]
```diff
@@ -1,8 +1,20 @@
...
 
+SCHEMAS = {
+    'certificates': None,
+    'gfx': None,
+    'add-ons': None,
+    'plugins': None,
+}
+
+with codecs.open(main.SCHEMA_FILE, encoding='utf-8') as f:
+    SCHEMAS = json.load(f)
```
Comment (magopian) [inline xml2kinto/tests/test_main.py]: I don't see a test needing a real schema, so I would drop those two lines and use the fake/empty schemas from above.

**Comment by magopian** [comment]
Huge work, thanks @Natim !

A few comments, let me know what you think ;)

File: .travis.yml (hunk 1 above)
Comment (Natim) [inline .travis.yml]: > Or is it needed to have the before_install?

yes :)

File: Makefile (hunk 2 above)
Comment (Natim) [inline Makefile]: I named it like that because it is how we did elsewhere e.g [kinto-signer](https://github.com/Kinto/kinto-signer/blob/master/Makefile#L62-L63) and [kinto-emailer](https://github.com/Kinto/kinto-emailer/blob/add-tests/Makefile#L53-L54)

File: Makefile (hunk 3 above)
Comment (Natim) [inline Makefile]: So basically this line is making sure kinto is running before starting the functional tests. ([See here](https://github.com/mozilla-services/xml2kinto/pull/34/files#diff-b67911656ef5d18c4ae36cb6741b7965R45))

File: config/kinto.ini (hunk 5 above)
Comment (Natim) [inline config/kinto.ini]: You mean in [kinto.tpl](https://github.com/Kinto/kinto/blob/master/kinto/config/kinto.tpl)?

File: config/kinto.ini (hunk 6 above)
Comment (Natim) [inline config/kinto.ini]: Yes it is just a file generated from ``kinto init``

File: schemas.json (hunk 7 above)
Comment (Natim) [inline schemas.json]: Because the file is used for tests I add to add it, also the Makefile command could be seen as one to update this file if the remote changed.

However I believe the file will stop changing at some point.

File: tox.ini (hunk 4 above)
Comment (Natim) [inline tox.ini]: Usually I try to avoid having tox to call Make because they do not necessary use the same virtualenv. I would rather have ``make functional`` call ``tox -e functional`` what do you think?

File: xml2kinto/__main__.py (hunk 9 above)
Comment (Natim) [inline xml2kinto/__main__.py]: I believe yes.
As you can see I am also using it in tests.

File: xml2kinto/__main__.py (hunk 10 above)
Comment (Natim) [inline xml2kinto/__main__.py]: Actually it is relevant to the functional tests.
The scrapping can arm AMO and slow down functional tests by a lot. That's the reason why this is there.

File: xml2kinto/__main__.py (hunk 11 above)
Comment (Natim) [inline xml2kinto/__main__.py]: You are right it should be the collection in both case. The rational was to use the defined collection arg first and fallback to the default.

File: xml2kinto/tests/test_main.py (hunk 14 above)
Comment (Natim) [inline xml2kinto/tests/test_main.py]: Actually https://github.com/mozilla-services/xml2kinto/pull/34/files#diff-c17a4b480e20f1476567f04b1e1ec2f6R95 this is using it for all ``assert_arguments`` calls

**Comment by Natim** [comment]
I have updated the PR with your insights @magopian thanks.

**Comment by Natim** [comment]
The rebase and the functional tests showed a bug in one of the previous master that I fixed.

File: tox.ini (hunk 4 above)
Comment (magopian) [inline tox.ini]: that could also be a way to do it. It means that it'll install tox and create a venv for it, but why not. It makes sense to have both use the same "code" though, to make sure we don't introduce incoherences in the future, or experience a different behavior on travis and dev.

File: config/kinto.ini (hunk 6 above)
Comment (magopian) [inline config/kinto.ini]: ah, understood, it can stay that way then I guess.

File: xml2kinto/__main__.py (hunk 10 above)
Comment (magopian) [inline xml2kinto/__main__.py]: fair enough.

File: config/kinto.ini (hunk 6 above)
Comment (Natim) [inline config/kinto.ini]: I have shrink it down.

**Comment by magopian** [comment]
Thanks for the changes, I believe there's only https://github.com/mozilla-services/xml2kinto/pull/34#discussion_r56302328 left to fix, and we're all done!

r+wc
